avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/StatementInfo.java

```diff
@@ -0,0 +1,122 @@
+import java.sql.Statement;
+import java.util.Objects;
+
+/**
+ *  All we know about a statement. Encapsulates a {@link ResultSet}.
+ */
+public class StatementInfo {
+  private static final String NEXT_METHOD_NAME = "next";
+
+  final Statement statement; // sometimes a PreparedStatement
+  private ResultSet resultSet; // a Proxy, not the real ResultSet.
+  private long position = 0;
+
+  /**
+   * Support a Proxy against ResultSet in which <code>position</code> is incremented after
+   * every call to <code>next()</code>.
+   */
+  private class ResultSetInvocationHandler implements InvocationHandler {
+    private final ResultSet realResultSet;
+
+    private ResultSetInvocationHandler(ResultSet resultSet) {
+      this.realResultSet = resultSet;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      try {
+        Object o = method.invoke(realResultSet, args);
+        if (NEXT_METHOD_NAME.equals(method.getName())) {
+          position++;
+        }
+        return o;
+      } catch (InvocationTargetException ex) {
+        throw ex.getCause();
+      }
+    }
+
+  }
+
+  public StatementInfo(Statement statement) {
+    this.statement = Objects.requireNonNull(statement);
+  }
+
+  /**
+   * Set a ResultSet on this object.
+   * @param resultSet The current ResultSet
+   */
+  public void setResultSet(ResultSet resultSet) {
+    // Updates have null ResultSet from the Connection
+    if (null == resultSet) {
+      this.resultSet = null;
+      return;
+    }
+
+    Class<?> clz = resultSet.getClass();
+    // Create a Proxy around the ResultSet to let us increment `position` when `next()` is called
+    // but before the methods returns without re-implementing all of ResultSet.
+    this.resultSet = (ResultSet) Proxy.newProxyInstance(clz.getClassLoader(), clz.getInterfaces(),
+        new ResultSetInvocationHandler(Objects.requireNonNull(resultSet)));
+  }
+
+  /**
+   * @return The {@link ResultSet} for this Statement
+   */
+  public ResultSet getResultSet() {
+    return Objects.requireNonNull(this.resultSet);
+  }
+
+  /**
+   * @return True if the resultSet is not null, false otherwise.
+   */
+  public boolean isResultSetInitialized() {
+    return null != resultSet;
+  }
+
+  /**
+   * Consume <code>offset - position</code> elements from the {@link ResultSet}
+   * @param offset The offset to advance to
+   * @return True if the resultSet was advanced to the current point, false if insufficient rows
+   *      were present to advance to the requested offset.
+   */
+  public boolean advanceResultSetToOffset(long offset) throws SQLException {
+    Preconditions.checkArgument(offset >= 0 && offset >= position, "Offset should be "
+        + " non-negative and not less than the current position. " + offset + ", " + position);
+    while (position < offset) {
+      // Position increment via the proxy wrapping next()
+      if (!resultSet.next()) {
```

Comment:
Hrm. Doesn't look like Phoenix likes this. I'll have to see if there's a way around this.

```
java.lang.RuntimeException: java.sql.SQLFeatureNotSupportedException
	at org.apache.calcite.avatica.jdbc.JdbcMeta.propagate(JdbcMeta.java:706)
	at org.apache.calcite.avatica.jdbc.JdbcMeta.syncResults(JdbcMeta.java:787)
	at org.apache.calcite.avatica.remote.LocalService.apply(LocalService.java:212)
	at org.apache.calcite.avatica.remote.Service$SyncResultsRequest.accept(Service.java:2032)
	at org.apache.calcite.avatica.remote.Service$SyncResultsRequest.accept(Service.java:1)
	at org.apache.calcite.avatica.remote.ProtobufHandler.apply(ProtobufHandler.java:52)
	at org.apache.calcite.avatica.server.AvaticaProtobufHandler.handle(AvaticaProtobufHandler.java:56)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:245)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLFeatureNotSupportedException
	at org.apache.phoenix.jdbc.PhoenixResultSet.relative(PhoenixResultSet.java:800)
	at org.apache.calcite.avatica.jdbc.StatementInfo.advanceResultSetToOffset(StatementInfo.java:99)
	at org.apache.calcite.avatica.jdbc.JdbcMeta.syncResults(JdbcMeta.java:781)
	... 14 more
```